File: src/csharp/Problems/MinRemoveToMakeValid.cs
//https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses/

namespace LeetCode.Problems;

public sealed class MinRemoveToMakeValid : ProblemBase
{
    [Theory]
    [ClassData(typeof(MinRemoveToMakeValid))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it => it.Param("lee(t(c)o)de)").Result("lee(t(c)o)de"))
          .Add(it => it.Param("a)b(c)d").Result("ab(c)d"))
          .Add(it => it.Param("a)b()c)d").Result("ab()cd"))
          .Add(it => it.Param("))((").Result(""));

    private string Solution(string s)
    {
        var stack = new Stack<int>();

        var result = new StringBuilder();
        foreach (var ch in s)
        {
            result.Append(ch);
            switch (ch)
            {
                case '(':
                    stack.Push(result.Length - 1);
                    break;
                case ')' when stack.Count > 0:
                    stack.Pop();
                    break;
                case ')':
                    result.Remove(result.Length - 1, 1);
                    break;
            }
        }

        while (stack.Any())
        {
            result.Remove(stack.Pop(), 1);
        }

        return result.ToString();
    }
}

File: src/python/problems/MinRemoveToMakeValid.py
# https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses/
class MinRemoveToMakeValid:
    def Solution(self, s: str) -> str:
        stack = []
        result = list(s)
        for i, ch in enumerate(s):
            if ch == '(':
                stack.append(i)
            elif ch == ')' and len(stack) > 0:
                stack.pop()
            elif ch == ')':
                result[i] = ''

        while stack:
            result[stack.pop()] = ''

        return ''.join(result)


MinRemoveToMakeValid().Solution("babad")
